Add Type.Id

dbuenzli opened the pull request:
As [promised](https://github.com/ocaml/ocaml/issues/11564#issuecomment-1265160760) this PR adds type identifiers to the stdlib.

dbuenzli committed:
Type.eq: fix doc string attachement.

dbuenzli committed:
Add Type.Id

yallop commented on stdlib/type.ml:
```diff
@@ -13,4 +13,31 @@
+
+  type 'a t = (module ID with type t = 'a)
+
+  let make () (type s) =
+    let module T = struct
+      type t = s
+      type _ id += Id : t id end
+    in
+    (module T : ID with type t = s)
```
```suggestion
  let make () (type s) : s t =
    (module struct
       type t = s
       type _ id += Id : t id end)

```

yallop commented on stdlib/type.ml:
```diff
@@ -13,4 +13,31 @@
+    in
+    (module T : ID with type t = s)
+
+  let equal (type t0) (type t1) (t0 : t0 t) (t1 : t1 t) : (t0, t1) eq option
+    =
+    let module T0 = (val t0 : ID with type t = t0) in
+    let module T1 = (val t1 : ID with type t = t1) in
+    match T0.Id with T1.Id -> Some Equal | _ -> None
```
```suggestion
  let equal (type a b) ((module A): a t) ((module B): b t) : (a, b) eq option
    = match A.Id with B.Id -> Some Equal | _ -> None
```

yallop reviewed (commented):
I'm in favour of this change.  I use something like this quite often.

I've suggested a couple of ways to make the implementation more succinct.

gasche commented:
My own implementation uses a bespoke datatype instead of `('a, 'b) eq option` to avoid the indirection:

```ocaml
type (_, _) ty_eq_test =
  | Eq : ('a, 'a) ty_eq_test
  | Unknown : ('a, 'b) ty_eq_test
```

I wonder whether this is something that should be considered here, for performance reasons.

dbuenzli committed:
Integrate improvements from @yallop.

dbuenzli committed:
Pre-allocate Some Equal, towards @gasche suggestion.

lthls commented:
> My own implementation uses a bespoke datatype instead of `('a, 'b) eq option` to avoid the indirection:
> 
> ```ocaml
> type (_, _) ty_eq_test =
>   | Eq : ('a, 'a) ty_eq_test
>   | Unknown : ('a, 'b) ty_eq_test
> ```
> 
> I wonder whether this is something that should be considered here, for performance reasons.

In this particular example, I think performance is going to be very similar. In bytecode, I expect that you get an extra `Isint` operation when matching on the option version, but I wouldn't be surprised if in native code this ends up mostly equivalent, maybe even very slightly faster with the option version. (Since `eq` is a type with a single constructor, after dead code elimination `match x with | Some Eq -> e1 | None -> e2` is equivalent to `match x with | Some _ -> e1 | None -> e2`, so there's no actual indirection involved.)

So in my opinion you're free to pick the version that makes most sense from a design perspective.

dbuenzli commented:
I integrated @yallop's improvements (thanks I didn't even know anonymous modules were a thing!).

What do other people think of @gasche's suggestion ? 

1. From the performance perspective. As a first measure I pre-allocated `Some Equal`. ~~Is the resulting test still less efficient than what he propose ? Do we get a load ? or is the pattern matching code smart enough ?~~ (see @lthls's response). 
2. From a code reading perspective. While it feels a bit silly not to reuse `Type.Equal`, I like the explicit `Unknown`. Of course options are also nice to use with the `Option` module but I doubt it's that useful in that case.

dbuenzli committed:
Make that fit on a line.

gasche commented:
(Now that I think of it, `Some Equal` is going to be shared as an immutable constant, so even the "preallocation" logic may not be necessary.)

dbuenzli commented:
So basically the discussion is now: do we want to read (from the example from the docs): 

```ocaml
let find : type a. a key -> t -> a option = fun k d ->
  match List.assq_opt k.exist d with
  | None -> None
  | Some (B (k', v)) ->
      match Type.Id.equal k.tid k'.tid with
      | None -> None
      | Some Type.Equal -> Some v
```

or 

```ocaml 
let find : type a. a key -> t -> a option = fun k d ->
  match List.assq_opt k.exist d with
  | None -> None
  | Some (B (k', v)) ->
      match Type.Id.equal k.tid k'.tid with
      | Type.Id.Unknown -> None
      | Type.Id.Equal -> Some v
```

I don't have a strong opinion on it. 

On the plus side with @gasche's proposal is that it embodies in the code what I wrote in the doc preamble (namely that unequal identifiers does not imply unequal types). On the minus side it's more definitions/explanations (though I can redirect to the explanation of `Type.eq`).

A twist could be to find another name for `equal` so that `None` and `Some Equal` also embody the idea of "don't know" and "equal". I tried but found nothing convincing (e.g. `equality_proof`), `base` uses `same_witness` which I'm not very fond of.

wikku commented on stdlib/type.mli:
```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
+  (** {1:ids Type identifiers} *)
+
+  type 'a t
+  (** The type for type identifiers. ['a] is the represented type. *)
```
`represented` might be confusing if `Type.Repr` [is added](https://github.com/ocaml/ocaml/issues/11564#issuecomment-1261396406)

dbuenzli replied:
This is consistent with the terminology introduced in the preamble. I personally wouldn't be confused by it. But if you think so, the please provide a full terminological story.

wikku replied:
I was thinking of having `to prove they identify the same type`, `a type can have multiple identifiers`, `the type of identifiers for ['a]`, but now I think having `represent` is good so as not to sound tautological.

dbuenzli replied:
If the goal is to avoid *represent* maybe we could use *denote*. What about:

```
A type identifier is a value that denotes a type. Given two 
type identifiers they can be tested for {{!Id.equal}equality} 
to prove they denote the same type. Note that unequal identifiers
does not imply unequal types: a given type can be denoted by more
than one identifier.

type 'a t
(** The type for identifiers for type ['a]. *)
```

wikku commented on stdlib/type.mli:
```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
+  (** {1:example Example}
+
+      The following shows how type identifiers can be used to implement
+      an heterogeneous key-value dictionary.
```
```suggestion
      a heterogeneous key-value dictionary.
```

wikku commented on stdlib/type.mli:
```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
+
+  let key () = let rec k = { tid = Type.Id.make (); exist = K k } in k
+
+  type binding = B : 'a key * 'a -> binding
```
Slight simplification (if the one I suggested above isn't applied):

```suggestion
  type binding = B : 'a Type.Id.t * 'a -> binding
```

(construct `B (k.tid, v)` in `add`, use the matched `tid` directly in `find`)

dbuenzli replied:
I prefer to keep the full key here. This makes the example easier to tweak with more interesting functionality like augmenting `'a key` with a name and a pretty-printer which in turns allows to provide a formatter for dictionaries.

wikku commented on stdlib/type.mli:
```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
+  val find : 'a key -> t -> 'a option
+  (** [find k d] is the binding of [k] in [d], if any. *)
+end = struct
+  type 'a key = { tid : 'a Type.Id.t; exist : exist_key }
```
I think we can simplify the example (avoid the unrelated issues of recursive values and physical equality) with this type:

```suggestion
  type 'a key = 'a Type.Id.t
```

(Wrap the key in `K` when calling list functions and use `assoc` instead of `assq`. This is irrelevant to the example, but this won't even allocate if `K` is `[@@unboxed]` .)

dbuenzli replied:
I'm uneasy with calling polymorphic comparison on first class modules. What do other people think ?

dbuenzli replied:
Besides this shows how to avoid needless allocations on each dictionary operation. I think I rather keep it that way.

dbuenzli replied:
And as mentioned above I think the example is more interesting that way. If people think like @wiktorkuchta, speak up.

wikku commented on stdlib/type.mli:
```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
+    | None -> None
+    | Some (B (k', v)) ->
+        match Type.Id.equal k.tid k'.tid with
+        | None -> None
```
This could be clearer if marked as an internal error.

dbuenzli committed:
Address some of @wiktorkutcha's comments.

code-ghalib commented on stdlib/type.mli:
```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
+
+  (** {1:example Example}
+
+      The following shows how type identifiers can be used to implement
```
Sorry, I'm having trouble understanding this example, and maybe my sources of confusion can suggest some improvements:
- What is a "heterogenous dictionary" , when is it useful? This signature seems a bit useless to me right now - I still have to collect the keys somewhere to be able to look things up, and the keys are of different types... so do I need another heterogenous collection somewhere? (Possibly this is motivated by [hmap](https://erratique.ch/software/hmap/doc/Hmap.html), maybe exposing the `Key.hide_type` is essential for this to be useful?) 
- Typically dictionary "keys" are not opaque things you have to create/concoct (e.g. words -> definitions, you already know the words), is dictionary the right name to use for this?
- The `key` function is really confusing - you're creating something (`'a t`) out of nothing (`unit`) using a cyclical value that can be of any type you cast it to (because memory representations are equivalent?) - it's not obvious that the comparisons in `remov_assq` and `assq_opt` are halal for this cyclical value - at least not until you look at their documentation or implementation to see that they use physical equality and not structural. Is there perhaps a less nifty, more beginner-friendly version of this? Even the `Type.Id.make` implementation was easier to understand for me than this, magic-module syntax notwithstanding.

gasche replied:
An example of "heterogeneous dictionary" is a record, with a fixed number of fields (keys) that can access values of different types.

dbuenzli replied:
> * What is a "heterogenous dictionary" , when is it useful?

Yes, I'll add a few words about that.

> Possibly this is motivated by [hmap](https://erratique.ch/software/hmap/doc/Hmap.html), maybe exposing the Key.hide_type is essential for this to be useful?)

No, in fact once you hide the type, it becomes less useful (unless you hide it in conjunction with something interesting like a function acting on the hidden type).  

> * Typically dictionary "keys" are not opaque things you have to create/concoct (e.g. words -> definitions, you already know the words), is dictionary the right name to use for this?

There are various names for structures that allow you to map key to values, hash tables, maps, [property lists](http://mlton.org/PropertyList), dictionaries, key-value stores. I just took a name that is not used by the stdlib and avoids fights about module casing conventions :–) If you want something else propose a better name. `Store` could be one but I think I rather keep it that way.

> * it's not obvious that the comparisons in `remov_assq` and `assq_opt` are halal for this cyclical value - at least not until you look at their documentation or implementation to see that they use physical equality and not structural

There's not much I can do about the name of these function which are likely inspired by lisp conventions. I notice though that, "the name is weird let's look up the docs" worked here :–)

wikku replied:
This structure reminded me of tickets as in "cloakroom ticket", so that could be another name for keys. This indicates they are created by the structure instead of chosen by the client and the name is already used in concurrency algorithms: ticket locks. I think "store" would be good.

If `'a key` is replaced with `'a Type.Id.t` similarly to my previous suggestion, it is simply an association list associating values of different types with their type identifiers.

code-ghalib replied:
> There's not much I can do about the name of these function which are likely inspired by lisp conventions. I notice though that, "the name is weird let's look up the docs" worked here :–)  

It wasn't the names that had me confused - I had not encountered cyclical values before, and had to read [this answer](https://stackoverflow.com/a/44369585) first to understand a little more what was going on. Also, maybe some explanation around `exist_key` would be helpful? For example, why can't you just use `tid` to find things in the list, why do you need the `exist_key`?  

fwiw, I think it's also a fair position, to say that if you don't understand this part of the `Stdlib`, then don't worry about it - fine line to walk between api documentation and full-fledged blog post.

Also, do we need a note about not marshalling these? [This answer](https://discuss.ocaml.org/t/types-as-first-class-citizens-in-ocaml/2030/3) which references [this library](https://github.com/let-def/distwit) seems to suggest so.

dbuenzli replied:
>  Also, maybe some explanation around exist_key would be helpful? For example, why can't you just use tid to find things in the list, why do you need the exist_key?

That's for the compiler to tell you :-) 

In any case if we add @hhugo's suggestion we can simplify the example and use regular assoc lists.

> fwiw, I think it's also a fair position, to say that if you don't understand this part of the Stdlib, then don't worry about it

My take is a bit different: here's a (possibly confusing) example, spend a bit of quality time fiddling with it, you could end up learning quite a bit from it – as you did.

hhugo commented:
I think it would be useful to also provide a way to get a "unique id" from a `_ Id.t`. It could be used to implement a `Univ_map`.

`Base` has the following.
```
    let uid (type a) (module M : S with type t = a) =
      Caml.Obj.Extension_constructor.id (Caml.Obj.Extension_constructor.of_val M.Key)
```

dbuenzli commented:
> I think it would be useful to also provide a way to get a "unique id" from a `_ Id.t`. It could be used to implement a `Univ_map`.

Indeed. I usually separately generate uids for my keys with a:

```ocaml 
let make_id =
  let id = Atomic.make 0 in fun () -> Atomic.fetch_and_add id 1
```

since I'm very religious about not using `Obj`. But if we can get it for free I'm happy to have it. So if I get a green light from the compiler devs to use the code you suggest I'm happy to add it.

gasche commented:
`Extension_constructor.of_val` is in Obj because it can only soundly be called on values that we know are extension constructors. In this example it is called with `M.Key`, which is an extension constructor, so indeed this is safe.

dbuenzli committed:
Remove pre-allocation of Some Equal. The compiler is good enough.

dbuenzli committed:
Code formatting consistency.

dbuenzli committed:
Add Type.Id.uid as per @hhugo's wish.

dbuenzli committed:
Improve docs, redo example with Type.Id.uid.

dbuenzli commented:
So I pushed a series of commit that:

0. Remove the preallocation of `Some Equal`, the compiler is good enough.
1. Added `Type.Id.uid` as per @hhugo's wish.
2. Reworked the docs to avoid mentions of `represent`.
3. Added a note about marshaling these values.
4. Simplified the example to use `Type.Id.uid`.

Regarding the so far undiscussed suggestion of @gasche I think I'm now leading towards keeping the option on the ground of definitional economy (but still open on disenting views).

dbuenzli committed:
Adjust stdlib/.depend.

dbuenzli committed:
Fix stdlib link order.

nojb commented on stdlib/type.mli:
```diff
@@ -35,3 +35,86 @@ type (_, _) eq = Equal: ('a, 'a) eq
+(** Type identifiers.
+
+    A type identifier is a value that denotes a type. Given two type
+    identifiers they can be tested for {{!Id.equal}equality} to prove
```
```suggestion
    identifiers, they can be tested for {{!Id.equal}equality} to prove
```

nojb commented on stdlib/type.mli:
```diff
@@ -35,3 +35,86 @@ type (_, _) eq = Equal: ('a, 'a) eq
+    identifiers they can be tested for {{!Id.equal}equality} to prove
+    they denote the same type. Note that:
+
+    - Unequal identifiers does not imply unequal types: a given type can be
```
```suggestion
    - Unequal identifiers do not imply unequal types: a given type can be
```

nojb commented on stdlib/type.mli:
```diff
@@ -35,3 +35,86 @@ type (_, _) eq = Equal: ('a, 'a) eq
+  (** {1:example Example}
+
+      The following shows how type identifiers can be used to
+      implement a simple heterogeneous key-value dictionary which you
+      can see as an extensible record type. In contrast to
```
Nitpicking, but personally I would remove `which you can see as an extensible record type`; I feel it makes the explanation a bit more complicated than it could be...
```suggestion
      implement a simple heterogeneous key-value dictionary. In contrast to
```

nojb reviewed (commented):
LGTM

What do you think of adding `Type.Id.compare`, `Type.Id.equal`, `Type.Id.hash` and `Type.Id.seeded_hash` (re-exposing the same operation on `uid`)? That would allow easily making hash tables, maps, etc. with type identifiers as keys.

gasche commented:
> That would allow easily making hash tables, maps, etc. with type identifiers as keys.

I've never used a Type.Id to index a map or hashtable -- my (somewhat common) is to store alongside a value in heterogeneous containers. When do you need that?

dbuenzli commented:
> What do you think of adding `Type.Id.compare`, `Type.Id.equal`, `Type.Id.hash` and `Type.Id.seeded_hash` (re-exposing the same operation on `uid`)?

If we do that, which new name do you suggest for `Type.Id.equal` as it exist ?

nojb commented:
> > That would allow easily making hash tables, maps, etc. with type identifiers as keys.
> 
> I've never used a Type.Id to index a map or hashtable -- my (somewhat common) is to store alongside a value in heterogeneous containers. When do you need that?

In the example given in the documentation of the present PR, the `uid` of type identifiers are used as keys of the heterogenous dictionary (implemented as an association list). In my mind, this is equivalent to using the type identifier itself as the key. But perhaps it is better to be explicit about the distinction between the two objects?

nojb commented:
> > What do you think of adding `Type.Id.compare`, `Type.Id.equal`, `Type.Id.hash` and `Type.Id.seeded_hash` (re-exposing the same operation on `uid`)?
> 
> If we do that, which new name do you suggest for `Type.Id.equal` as it exist ?

Good point. Before embarking on a naming discussion, what's your take on the question of adding these functions? (To be clear: I do not have much experience with type identifiers, so perhaps my suggestion was misguided; opinions welcome.)

dbuenzli commented:
> Before embarking on a naming discussion, what's your take on the question of adding these functions?

I think it's always nice if the datatypes provided by the stdlib implement `Map.Ord` and/or `Hashtbl.HashedType`. 

But in this case we can't since `Type.Id.t` is polymorphic. Unless I'm mistaken you won't be able to use them the way you thought you could because the type system will say: no! 

So I think adding them will be mostly useless, except maybe for:

1. A hypothetic deriving mecanism or for quick off-the-shelf answers for people writing corresponding functions for a larger type that includes them.

For the usage you likely intented to, they would for sure be nice to have on an existential of `Type.Id.t`. Which we could have in a submodule if people want to, but:

2. We open yet another can of interesting naming conventions to establish in the stdlib. Name of the module, name of the existential constructor, e.g. `Type.Id.Exist.V` ? `Type.Id.Untyped.E[xist]` ? 
3. My impression is that in general you want the existential to pack more than just the type identifier, for example a codec or printer for the type. So that `Stdlib` provided existential module would end up being little used.

I started writing this message convinced that they wouldn't be useful but then got to 1. at some point. I'm no longer sure :-)

dbuenzli committed:
Make the module more conventional and address @nojb's review.

dbuenzli commented:
So I renamed `equal` to `typed_equal` and on the ground of "better safe than sorry" and "don't make me think" made the module follow standard conventions – even though that doesn't turn it into a functor argument for `{Set,Map,Hashtbl}.Make`.

nojb commented:
Where are we in this PR? I feel I may have derailed an unusually consensual PR with my remark about the `compare`, `equal`, `hash`, ...

From a distance, the PR looks fine as it is, but I prefer not to be the first approval as I don't have a lot of experience with type identifiers.

If people are still in two minds about the addition of these functions, perhaps it is better to remove them, but keeping `typed_equal` instead of `equal` so that they could be added later, if needed.

yallop commented:
> If people are still in two minds about the addition of these functions, perhaps it is better to remove them, but keeping `typed_equal` instead of `equal` so that they could be added later, if needed.

That'd be my preference.  I'm uneasy about the fact that those functions don't match the standard signatures, and I don't see a strong need for including them, since having `uid` makes it easy to implement them externally.

(On the name "`typed_equal`": I'd prefer something with an adverb in place of `typed` ─ e.g. `match provably_equal x y with ...` ─ but I don't want to derail the PR by starting a naming battle.)

yallop commented on stdlib/type.mli:
```diff
@@ -35,3 +35,100 @@ type (_, _) eq = Equal: ('a, 'a) eq
+
+  (** {1:ids Type identifiers} *)
+
+  type 'a t
```
```suggestion
  type !'a t
```
(Injectivity annotations for type constructors that are necessarily injective are now [standard style in the standard library](https://github.com/ocaml/ocaml/pull/9781))

dbuenzli replied:
The type system thinks this is invariant: 

```
29 |   type !'a t = (module ID with type t = 'a)
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: In this definition, expected parameter variances are not satisfied.
       The 1st type parameter was expected to be injective invariant,
       but it is invariant.
```

yallop replied:
That's interesting.  The injectivity annotation was accepted in OCaml 5.0.0.

yallop replied:
See #11869.  There's no reason for the issue to block the merge of this PR; we can always add an injectivity annotation later if necessary.

dbuenzli committed:
Remove standard non-standard functions.

dbuenzli committed:
Rename type_equal to provably_equal.

nojb commented:
To be merged, this PR needs two official approvals. @yallop, @gasche: you have already reviewed the PR, would you like to formally approve it?

jhwoodyatt commented:
I think the type parameter of a type identifier can be injective.

Stupid question: is there a good reason the type parameter of a type identifier has to be invariant? Seems like it would be better if it were covariant, but I don't think the type checker allows that.

jhwoodyatt commented on stdlib/type.mli:
```diff
@@ -35,3 +35,85 @@ type (_, _) eq = Equal: ('a, 'a) eq
+
+  (** {1:ids Type identifiers} *)
+
+  type 'a t
```
Seems like this could be `type !'a t` and that might help?

dbuenzli replied:
Indeed, but see the resolved discussion above.

yallop commented:
> Stupid question: is there a good reason the type parameter of a type identifier has to be invariant? Seems like it would be better if it were covariant, but I don't think the type checker allows that.

It'd be unsound for it to be covariant.  Covariance would allow you to create an id of type `<m:int> Type.Id.t`, cast it to an id of type `< > Type.Id.t`, use `provably_equal` to obtain a proof that `<m:int>` and `< >` were equal, then use the proof to access the non-existent `m` method of an empty object.

jhwoodyatt commented:
> It'd be unsound for it to be covariant.

Yeah, I had a hunch it was a stupid question. I think what I want is a similar way to prove that a type is a subtype of another type.

gasche commented:
The injectivity-inference issue was fixed in trunk ( #11874 ) so the type could probably be annotated injective now.

gasche commented:
@fpottier remarked to me tha (as is well-known) t heterogeneous datatypes can be implemented with a universal type instead of a type identifier.

```ocaml
module Univ : sig
  (* A universal type. *)

  type t

  type 'a embedding = ('a -> t) * (t -> 'a)
  (* An embedding is an injection-projection pair ([inject], [project]),
     where [inject] injects a value of type ['a] into our universal type
     [t], and [project (inject x)] is [x].

     [project] raises the [Not_found] exception on universal values
     that were not build using the paired [inject] function. *)
                                  
  val make: unit -> 'a embedding
end = struct
  type t = ..

  type 'a embedding = ('a -> t) * (t -> 'a)
  
  let make (type a) () : a embedding =  
    let module T = struct
      type t += Tag of a
      let inject (x : a) : t =
        Tag x
      let project (u : t) : a =
        match u with
        | Tag x -> x
        | _ -> raise Not_found
    end in
    T.inject, T.project
end


module Dict : sig
  type t
  (** The type for dictionaries. *)
  type 'a key
  (** The type for keys binding values of type ['a]. *)
  val key : unit -> 'a key
  (** [key ()] is a new dictionary key. *)
  val empty : t
  (** [empty] is the empty dictionary. *)
  val add : 'a key -> 'a -> t -> t
  (** [add k v d] is [d] with [k] bound to [v]. *)
  val remove : 'a key -> t -> t
  (** [remove k d] is [d] with the last binding of [k] removed. *)
  val find : 'a key -> t -> 'a option
  (** [find k d] is the binding of [k] in [d], if any. *)
end = struct
  type 'a key = {
    embed : 'a Univ.embedding;
    key : int;
  }
  type t = (int * Univ.t) list

  let key () =
    let key = Oo.id object end in
    { embed = Univ.make (); key }

  let empty = []
  let add k v d = (k.key, fst k.embed v) :: d
  let remove k d = List.remove_assoc k.key d
  let find : type a. a key -> t -> a option = fun k d ->
    match List.assoc_opt k.key d with
    | None -> None
    | Some u ->
      match snd k.embed u with
      | exception Not_found -> assert false
      | v -> Some v
end
```

This implementation is arguably simpler (no GADTs).
I am not sure how to extend it to store "type metadata" (information on how to print, compare etc.) inside the data-structure. Doing it in the same style as the PR code would require adding another existential. It could also be done by extending the type Univ.embedding, but this solution seems less modular than with type tags, where each user can decide which metadata they want separately.

dbuenzli committed:
Add injectivity annotation.

yallop commented:
> The injectivity-inference issue was fixed in trunk ( #11874 ) so the type could probably be annotated injective now.

Yes.  I think there's a bootstrap needed first, though.

dbuenzli commented:
> Yes. I think there's a bootstrap needed first, though.

I does need yes. Can someone push one to trunk ?

yallop commented:
> Can someone push one to trunk ?

Pushed: f48cc5351b1cffe834b3e1a817cfae0f3dda2070

dbuenzli commented:
> This implementation is arguably simpler (no GADTs).

If we are on a battle to use the less features as possible then, [here is](https://github.com/dbuenzli/topkg/blob/e30927fed486e4642e9f167908fd92cbd44609d1/src/topkg_conf.ml#L50-L53) a `Univ` without (explicit) extensible variants
and [here](https://groups.google.com/g/fa.caml/c/iSlmXtF1bhI) is how to do an heterogenous dict without first class modules – one functor application to generate keys for a given type, much less convenient, sometimes I forget how much more constrained the type system was in the first decade of the century.

> I am not sure how to extend it to store "type metadata" (information on how to print, compare etc.) inside the data-structure.

I don't think it poses a big problem you just need to embed the key with it (and have the metadata in the key), i.e. tweak your example to, for example:

```ocaml
type 'a key = {
    embed : ('a key * 'a) Univ.embedding;
    key : int;
    pp : Format.formatter -> 'a -> unit
  }
```

nojb reviewed (approved):
LGTM

As it is the weekend, let's wait until Monday before merging to give a chance for any last-minute opinions to come up.

fpottier commented:
I suggest documenting whether it is safe/unsafe to serialize and deserialize these type identifiers.

dbuenzli commented:
It's already mentioned [here](https://github.com/ocaml/ocaml/pull/11830/files#diff-842f6612822584c5687d3d37be7c27d6dcd3e1d2c8df3d18f59e5258d4193c36R49-R50). What do you think is missing ?

fpottier commented:
> It's already mentioned here. What do you think is missing ?

Sorry, I missed that. Fine, then.

nojb commented:
Let's merge, thanks!
